### PR TITLE
Cpp win creation

### PR DESCRIPTION
**Pull request by RadekVana**

### CODE REVIEW

File: src/guiapi/src/window_icon.cpp [hunk 1]
```diff
@@ -9,35 +11,91 @@ void window_icon_init(window_icon_t *window) {
...
     id_res = id;
     Invalidate();
 }
+
+window_icon_t::window_icon_t(window_t *parent, rect_ui16_t rect, uint16_t id_res, is_closed_on_click_t close)
+    : window_t(parent, rect, is_dialog_t::no, close)
+    , id_res(id_res)
+    , alignment(ALIGN_CENTER) {
+}
+//Icon rect is increased by badding, than icon is centered inside it
```
Comment (espr14) [inline src/guiapi/src/window_icon.cpp:42]: Icon rect is increased by padding, than icon is centered inside it
```suggestion
//Icon rect is increased by padding, icon is centered inside it
```

File: src/guiapi/src/window_icon.cpp [hunk 2]
```diff
@@ -9,35 +11,91 @@ void window_icon_init(window_icon_t *window) {
...
+                ret.h = sz.h + padding.top + padding.bottom;
+            }
+            return ret;
+        }(),
+        id_res, close) {
+}
+
+void window_icon_t::unconditionalDraw() {
+    uint8_t ropfn = 0;
+    if ((IsBWSwapped())) { // that could not be set, but what if
```
Comment (espr14) [inline src/guiapi/src/window_icon.cpp:62]:
```suggestion
    if (IsBWSwapped()) { // that could not be set, but what if
```

File: src/guiapi/src/window_icon.cpp [hunk 3]
```diff
@@ -9,35 +11,91 @@ void window_icon_init(window_icon_t *window) {
...
+        ropfn |= ROPFN_DISABLE;
+    }
+    if ((IsFocused())) {
```
Comment (espr14) [inline src/guiapi/src/window_icon.cpp:65]:
```suggestion
    if (IsFocused()) {
```

File: src/guiapi/src/window_icon.cpp [hunk 4]
```diff
@@ -9,35 +11,91 @@ void window_icon_init(window_icon_t *window) {
...
+        ropfn |= ROPFN_SWAPBW;
+    }
+
+    render_icon_align(rect, id_res, color_back, RENDER_FLG(alignment, ropfn));
+}
+
+bool window_icon_t::IsBWSwapped() const { return f_parent_defined0 == true; }
```
Comment (espr14) [inline src/guiapi/src/window_icon.cpp:72]:
```suggestion
bool window_icon_t::IsBWSwapped() const { return f_parent_defined0; }
```

File: src/guiapi/src/window_frame.cpp [hunk 5]
```diff
@@ -1,92 +1,223 @@
...
+                //define shadowed by dialog flag
+                pWin = pWin->GetNext();
+            }
+        }
+        last->SetNext(win);
+        last = last->GetNext();
+    } else {
+        first = last = win;
+    }
 }
```
Comment (espr14) [inline src/guiapi/src/window_frame.cpp:30]:
```suggestion
void window_frame_t::RegisterSubWin(window_t *win) {
    if (!first || !last) {
        first = last = win;
        return;
    }        
    if (win->IsDialog()) {
        window_t *pWin = first;
        while (pWin) {
            pWin->Hide(); //todo check dialog intersection
            //define shadowed by dialog flag
            pWin = pWin->GetNext();
        }
    }
    last->SetNext(win);
    last = last->GetNext();
}
```

File: src/guiapi/src/window_frame.cpp [hunk 6]
```diff
@@ -1,92 +1,223 @@
...
+}
+
+window_t *window_frame_t::GetLast() const {
+    return last;
+}
+
+void window_frame_t::draw() {
+    if (!IsVisible())
+        return;
+    bool setChildernInvalid = false;
```
Comment (espr14) [inline src/guiapi/src/window_frame.cpp:71]:
```suggestion
    bool setChildrenInvalid = false;
```

File: src/guiapi/src/window_frame.cpp [hunk 7]
```diff
@@ -1,92 +1,223 @@
...
+    while (ptr) {
+        if (setChildernInvalid)
+            ptr->Invalidate();
+        ptr->Draw();
+        ptr = ptr->GetNext();
+    }
+}
+
+void window_frame_t::windowEvent(window_t *sender, uint8_t event, void *param) {
+    int dif = (int)param;
```
Comment (espr14) [inline src/guiapi/src/window_frame.cpp:89]: (int)*param?

File: src/guiapi/src/window_frame.cpp [hunk 8]
```diff
@@ -1,92 +1,223 @@
...
-        id = window_focused();
-        while (dif--) {
-            id = window_prev_enabled(id);
+        while (pWin && dif--) {
+            window_t *const pPrev = GetPrevEnabledSubWin(pWin);
+            if (pPrev) {
+                pWin = pPrev;
+            } else {
+                break;
+            }
```
Comment (espr14) [inline src/guiapi/src/window_frame.cpp:106]:
```suggestion
            if (!pPrev) {
                break;
            pWin = pPrev;
```

File: src/guiapi/src/window_frame.cpp [hunk 9]
```diff
@@ -1,92 +1,223 @@
...
+void window_frame_t::screenEvent(window_t *sender, uint8_t ev, void *param) {
+    window_t *ptr = first;
+    while (ptr) {
+        ptr->ScreenEvent(sender, ev, param);
+        ptr = ptr->GetNext();
+    }
+    windowEvent(this, ev, param);
+}
+
+//resend invalidate to all childern
```
Comment (espr14) [inline src/guiapi/src/window_frame.cpp:155]:
```suggestion
//resend invalidation to all children
```

File: src/guiapi/src/window_frame.cpp [hunk 10]
```diff
@@ -1,92 +1,223 @@
...
-    {
-        WINDOW_CLS_FRAME,
-        sizeof(window_frame_t),
-        (window_init_t *)window_frame_init,
-        (window_done_t *)window_frame_done,
-        (window_draw_t *)window_frame_draw,
-        (window_event_t *)window_frame_event,
-    },
-};
+//resend event to all childern
```
Comment (espr14) [inline src/guiapi/src/window_frame.cpp:145]:
```suggestion
//resend event to all children
```

File: src/guiapi/src/window_frame.cpp [hunk 11]
```diff
@@ -1,92 +1,223 @@
...
+//resend invalidate to all childern
+void window_frame_t::invalidate(rect_ui16_t validation_rect) {
+    window_t *ptr = first;
+    while (ptr) {
+        ptr->Invalidate(validation_rect);
+        ptr = ptr->GetNext();
+    }
+}
+
+//resend validate to all childern
```
Comment (espr14) [inline src/guiapi/src/window_frame.cpp:164]:
```suggestion
//resend validation to all children
```

File: src/guiapi/src/window.cpp [hunk 12]
```diff
@@ -1,351 +1,215 @@
...
-        if (id < WINDOW_MAX_WINDOWS) { //id is valid
-            windows[id] = window;      //set window pointer
-            window_count++;            //increment count
-        } else
-            id = -1;
-    }
-    return id;
+bool window_t::IsVisible() const { return f_visible == true; }
+bool window_t::IsEnabled() const { return f_enabled == true; }
+bool window_t::IsInvalid() const { return f_invalid == true; }
```
Comment (espr14) [inline src/guiapi/src/window.cpp:12]:
```suggestion
bool window_t::IsVisible() const { return f_visible; }
bool window_t::IsEnabled() const { return f_enabled; }
bool window_t::IsInvalid() const { return f_invalid; }
```

File: src/guiapi/src/window.cpp [hunk 13]
```diff
@@ -1,351 +1,215 @@
...
+bool window_t::IsFocused() const { return GetFocusedWindow() == this; }
+bool window_t::IsCaptured() const { return GetCapturedWindow() == this; }
+bool window_t::HasTimer() const { return f_timer == true; }
```
Comment (espr14) [inline src/guiapi/src/window.cpp:15]:
```suggestion
bool window_t::HasTimer() const { return f_timer; }
```

File: src/guiapi/src/window.cpp [hunk 14]
```diff
@@ -1,351 +1,215 @@
...
 }
 
-window_class_t *class_ptr(int16_t cls_id) {
-    window_class_t *cls = 0;
-    if ((cls_id >= 0) && (cls_id < window_class_count) && ((cls = (window_class_t *)window_classes[cls_id]) != 0))
-        return cls;
-    if ((cls_id >= WINDOW_CLS_USER) && (cls_id < (WINDOW_CLS_USER + window_user_class_count)) && ((cls = window_user_classes[cls_id - WINDOW_CLS_USER]) != 0))
-        return cls;
-    return 0;
+//frame will invalidate childern
```
Comment (espr14) [inline src/guiapi/src/window.cpp:31]:
```suggestion
//frame will invalidate children
```

File: src/guiapi/src/window.cpp [hunk 15]
```diff
@@ -1,351 +1,215 @@
...
+void window_t::invalidate(rect_ui16_t validation_rect) {
 }
 
-window_t *window_ptr(int16_t id) {
-    return ((id >= 0) && (id < WINDOW_MAX_WINDOWS)) ? windows[id] : 0;
+//frame will validate childern
```
Comment (espr14) [inline src/guiapi/src/window.cpp:35]:
```suggestion
//frame will validate children
```

File: src/guiapi/src/window.cpp [hunk 16]
```diff
@@ -1,351 +1,215 @@
...
-        if (window == window_focused_ptr)
-            window_focused_ptr = 0;
-        if (window == window_popup_ptr)
-            window_popup_ptr = 0;
-        //if (window == window_0) window_0 = 0;
-        if (count == 0)
-            window_0 = 0;
+void window_t::Show() {
+    if (!IsVisible()) {
+        f_visible = 1;
```
Comment (espr14) [inline src/guiapi/src/window.cpp:72]:
```suggestion
        f_visible = true;
```

File: src/guiapi/src/window.cpp [hunk 17]
```diff
@@ -1,351 +1,215 @@
...
 
-void window_destroy_children(int16_t id) {
-    window_t *window;
-    int16_t id_child;
-    for (id_child = 0; id_child < WINDOW_MAX_WINDOWS; id_child++)
-        if (((window = windows[id_child]) != 0) && (window->id_parent == id))
-            window_destroy(id_child);
+void window_t::Hide() {
+    if (IsVisible()) {
+        f_visible = 0;
```
Comment (espr14) [inline src/guiapi/src/window.cpp:79]:
```suggestion
        f_visible = false;
```

File: src/guiapi/src/window.cpp [hunk 18]
```diff
@@ -1,351 +1,215 @@
...
-                        window->cls->draw(window);
-            } else if (window->cls->draw)
-                window->cls->draw(window);
-        }
+window_t *window_t::GetNextEnabled() const {
+    if (next) {
+        return (next->IsEnabled()) ? next : next->GetNextEnabled();
+    } else {
+        return nullptr;
+    }
```
Comment (espr14) [inline src/guiapi/src/window.cpp:146]:
```suggestion
    if (next)
        return (next->IsEnabled()) ? next : next->GetNextEnabled();
    return nullptr;
```

File: src/guiapi/src/window.cpp [hunk 19]
```diff
@@ -1,351 +1,215 @@
...
+    screenEvent(sender, ev, param);
 }
 
-uint16_t window_get_icon_id(int16_t id) {
-    window_t *window;
-    if ((window = window_ptr(id)) != 0) {
-        switch (window->cls->cls_id) {
-        case WINDOW_CLS_ICON:
-            return ((window_icon_t *)window)->id_res;
+//frame does something else - resends to all childern
```
Comment (espr14) [inline src/guiapi/src/window.cpp:190]:
```suggestion
//frame does something else - resend to all children
```

File: src/guiapi/src/window.cpp [hunk 20]
```diff
@@ -1,351 +1,215 @@
...
-        Invalidate();
-    }
+void window_t::WindowEvent(window_t *sender, uint8_t ev, void *param) {
+    windowEvent(sender, ev, param);
 }
 
-void window_t::SetBackColor(color_t clr) {
-    color_back = clr;
-    Invalidate();
+void window_t::ScreenEvent(window_t *sender, uint8_t ev, void *param) {
```
Comment (espr14) [inline src/guiapi/src/window.cpp:186]: Same method name with different capital letters? This is dangerous or useless.

File: src/guiapi/src/window_frame.cpp (hunk 7 above)
Comment (RadekVana) [reply src/guiapi/src/window_frame.cpp:89]: it is old code, to be removed.
there will be no void* parameter in future

File: src/guiapi/src/window.cpp (hunk 20 above)
Comment (RadekVana) [reply src/guiapi/src/window.cpp:184]: Public methods have Big capitals
they are usually not virtual and call protected virtual method

File: src/gui/dialogs/DialogRadioButton.cpp [hunk 21]
```diff
@@ -51,21 +53,28 @@ RadioButton &RadioButton::operator++() {
...
+        switch (event) {
+        case WINDOW_EVENT_ENC_UP:
+            ++(*this);
+            return;
+        case WINDOW_EVENT_ENC_DN:
+            --(*this);
+            return;
+        default:
+            window_t::windowEvent(sender, event, param);
+        }
```
Comment (espr14) [inline src/gui/dialogs/DialogRadioButton.cpp:74]:
```suggestion
    if (!GetParent())
        return;
        
    switch (event) {
    case WINDOW_EVENT_ENC_UP:
        ++(*this);
        return;
    case WINDOW_EVENT_ENC_DN:
        --(*this);
        return;
    default:
        window_t::windowEvent(sender, event, param);
    }
```

File: src/gui/ScreenHandler.cpp [hunk 22]
```diff
@@ -0,0 +1,203 @@
...
+        bsod("Screen stack overflow");
+    }
+}
+
+void Screens::Loop() {
+    if (close_all) {
+        if (current) {                              // is there something to close?
+            if (creator) {                          // have creator, have to emulate opening
+                stack_iterator = stack.begin();     // point to screen[0], (screen[0] is home)
+                close = false;                      // clr close flag, creator will be pusthed into screen[1] position
```
Comment (espr14) [inline src/gui/ScreenHandler.cpp:159]:
```suggestion
                close = false;                      // clr close flag, creator will be pushed into screen[1] position
```

File: src/gui/ScreenHandler.cpp [hunk 23]
```diff
@@ -0,0 +1,203 @@
...
+                close = true;                       // set flag to close screen[1] == open screen[0] (home)
+            }
+        }
+        close_all = false;
+    }
+
+    //open new screen
+    if (creator || close) {
+        if (current) {
+            current.reset(); //without reset screens does not behave correctly, I do not know why
```
Comment (espr14) [inline src/gui/ScreenHandler.cpp:171]:
```suggestion
            current.reset(); //without a reset screens do not behave correctly, I do not know why
```

File: src/gui/ScreenHandler.cpp [hunk 24]
```diff
@@ -0,0 +1,203 @@
...
+                    --stack_iterator;          // point to previous screen - will become "behind last creator"
+                    creator = *stack_iterator; // use previous screen as current creator
+                    close = false;
+                } else {
+                    bsod("Screen stack underflow");
+                }
+            } else {
+                if (stack_iterator != stack.end() && (stack_iterator + 1) != stack.end()) {
+                    ++stack_iterator;            // point behind last creator
+                    (*stack_iterator) = creator; // save future creator 0n top of the stack
```
Comment (espr14) [inline src/gui/ScreenHandler.cpp:183]:
```suggestion
                    (*stack_iterator) = creator; // save future creator on top of the stack
```

File: src/guiapi/src/window_frame.cpp (hunk 6 above)
Comment (RadekVana) [reply src/guiapi/src/window_frame.cpp:71]: fix

**Pull request merged**